test/boost/network_topology_strategy_test.cc

```diff
@@ -480,6 +530,148 @@ SEASTAR_THREAD_TEST_CASE(NetworkTopologyStrategy_tablets_test) {
     }
 }
 
+SEASTAR_THREAD_TEST_CASE(NetworkTopologyStrategy_tablet_allocation_balancing_test) {
+    auto my_address = gms::inet_address("localhost");
+
+    // Create the RackInferringSnitch
+    snitch_config cfg;
+    cfg.listen_address = my_address;
+    cfg.broadcast_address = my_address;
+    cfg.name = "RackInferringSnitch";
+    sharded<snitch_ptr> snitch;
+    snitch.start(cfg).get();
+    auto stop_snitch = defer([&snitch] { snitch.stop().get(); });
+    snitch.invoke_on_all(&snitch_ptr::start).get();
+
+    locator::token_metadata::config tm_cfg;
+    tm_cfg.topo_cfg.this_endpoint = my_address;
+    tm_cfg.topo_cfg.local_dc_rack = { snitch.local()->get_datacenter(), snitch.local()->get_rack() };
+
+    // Generate a random cluster
+    auto shard_count = 1 + tests::random::get_int(4);
+    std::vector<ring_point> ring_points;
+    std::vector<sstring> dcs;
+
+    double point = 1;
+    size_t num_dcs = 1;
+    size_t num_racks = 1 + tests::random::get_int(4);
+    size_t nodes_per_rack = 1 + tests::random::get_int(4);
+    size_t nodes_per_dc = num_racks * nodes_per_rack;
+    for (size_t dc = 0; dc < num_dcs; ++dc) {
+        sstring dc_name = fmt::format("{}", 100 + dc);
+        dcs.emplace_back(dc_name);
+        for (size_t rack = 0; rack < num_racks; ++rack) {
+            sstring rack_name = fmt::format("{}", 10 + rack);
+            for (size_t i = 1; i <= nodes_per_rack; ++i) {
+                ring_points.emplace_back(point, inet_address(format("192.{}.{}.{}", dc_name, rack_name, i)));
+            }
+        }
+    }
+
+    testlog.debug("num_dcs={} num_racks={} nodes_per_rack={} shards_per_node={} ring_points=[{}]", num_dcs, num_racks, nodes_per_rack, shard_count,
+            fmt::join(ring_points | boost::adaptors::transformed([] (const ring_point& rp) {
+                return fmt::format("({}, {})", rp.id, rp.host);
+            }), ", "));
+
+    // Initialize the token_metadata
+    locator::shared_token_metadata stm([] () noexcept { return db::schema_tables::hold_merge_lock(); }, tm_cfg);
+    stm.mutate_token_metadata([&] (token_metadata& tm) -> future<> {
+        auto& topo = tm.get_topology();
+        for (const auto& [ring_point, endpoint, id] : ring_points) {
+            std::unordered_set<token> tokens;
+            tokens.insert({dht::token::kind::key, d2t(ring_point / ring_points.size())});
+            topo.add_node(id, endpoint, make_endpoint_dc_rack(endpoint), locator::node::state::normal, shard_count);
+            tm.update_host_id(id, endpoint);
+            co_await tm.update_normal_tokens(std::move(tokens), id);
+        }
+    }).get();
+
+    auto tmptr = stm.get();
+    const auto& topo = tmptr->get_topology();
+
+    auto s = schema_builder("ks", "tb")
+        .with_column("pk", utf8_type, column_kind::partition_key)
+        .with_column("v", utf8_type)
+        .build();
+
+    for (auto test_case = 0; test_case < 3; test_case++) {
```

Comment:
Yikes, the "test_case" has some meaningful difference between cases, it's not an abstract "iteration with different random number(s)", right? In that case it's better to have several `NetworkTopologyStrategy_tablet_allocation_balancing_test_<something>` top-level cases calling this function with some argument that tweaks its behavior

Reply:
case 0 is rf:=1, case 2 us rf:=n and case 1, if n is large enough draws a random rf between 2 to n-1.

Reply:
in v4 I left on one random choice of rf_per_dc, and the random test function is called 100 times in a loop